hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithFSO.java

```diff
@@ -155,51 +177,75 @@ public boolean processWithFSO(OMUpdateEventBatch events) {
         return false;
       }
+
+      if (!checkOrphanDataAndCallWriteFlushToDB(orphanKeysMetaDataSetMap, 1L)) {
+        return false;
+      }
     }
 
     // flush and commit left out entries at end
     if (!flushAndCommitNSToDB(nsSummaryMap)) {
       return false;
     }
+    // flush and commit left out entries at end
+    if (!writeFlushAndCommitOrphanKeysMetaDataToDB(orphanKeysMetaDataSetMap,
+        1L)) {
+      return false;
+    }
 
     LOG.info("Completed a process run of NSSummaryTaskWithFSO");
     return true;
   }
 
   public boolean reprocessWithFSO(OMMetadataManager omMetadataManager) {
     Map<Long, NSSummary> nsSummaryMap = new HashMap<>();
+    Map<Long, OrphanKeyMetaData> orphanKeyMetaDataMap = new HashMap<>();
 
     try {
-      Table<String, OmDirectoryInfo> dirTable =
-          omMetadataManager.getDirectoryTable();
-      try (TableIterator<String,
-              ? extends Table.KeyValue<String, OmDirectoryInfo>>
-                dirTableIter = dirTable.iterator()) {
-        while (dirTableIter.hasNext()) {
-          Table.KeyValue<String, OmDirectoryInfo> kv = dirTableIter.next();
-          OmDirectoryInfo directoryInfo = kv.getValue();
-          handlePutDirEvent(directoryInfo, nsSummaryMap);
-          if (!checkAndCallFlushToDB(nsSummaryMap)) {
-            return false;
-          }
-        }
+      if (handleDirectoryTableEvents(omMetadataManager, nsSummaryMap,
+          orphanKeyMetaDataMap)) {
+        return false;
       }
 
-      // Get fileTable used by FSO
-      Table<String, OmKeyInfo> keyTable =
-          omMetadataManager.getFileTable();
-
-      try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
-              keyTableIter = keyTable.iterator()) {
-        while (keyTableIter.hasNext()) {
-          Table.KeyValue<String, OmKeyInfo> kv = keyTableIter.next();
-          OmKeyInfo keyInfo = kv.getValue();
-          handlePutKeyEvent(keyInfo, nsSummaryMap);
-          if (!checkAndCallFlushToDB(nsSummaryMap)) {
-            return false;
-          }
-        }
+      if (handleFileTableEvents(omMetadataManager, nsSummaryMap,
+          orphanKeyMetaDataMap)) {
+        return false;
+      }
+
+      if (!writeFlushAndCommitOrphanKeysMetaDataToDB(
+          orphanKeyMetaDataMap, 1L)) {
+        return false;
+      }
+
+      Set<Long> bucketObjIdSet = new HashSet<>();
+      buildBucketObjIdSet(omMetadataManager, bucketObjIdSet);
+
+      List<Long> orphanMetaDataKeyList = new ArrayList<>();
+      LOG.info("Starting to verify orphan key candidates...");
+      // If any deleted directory is present as parent key in
+      // orphanKeysMetaDataTable, then remove that parent directory entry
+      // from orphanKeysMetaDataTable, because child keys for deleted directory
+      // will not be treated as orphans.
+      Instant start = Instant.now();
+      if (removeDeletedDirEntries(omMetadataManager, orphanMetaDataKeyList)) {
```

Comment:
from method name, if success should return true, same for handleFileTableEvents, handleDirectoryTableEvents, removeDeletedDirEntries, verifyOrphanParentsForBucket returns true for failure case.

Reply:
Ok.